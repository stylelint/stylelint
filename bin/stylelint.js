#!/usr/bin/env node

'use strict';

// to use V8's code cache to speed up instantiation time
require('v8-compile-cache');

require('../lib/cli')(process.argv.slice(2));
