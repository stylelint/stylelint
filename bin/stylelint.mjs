#!/usr/bin/env node

import cli from '../lib/cli.mjs';

cli(process.argv.slice(2));
