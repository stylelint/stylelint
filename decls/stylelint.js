export type stylelint$configExtends = string | Array<string>;
export type stylelint$configPlugins = string | Array<string>;
export type stylelint$configProcessors =
  | string
  | Array<string | [string, Object]>;
export type stylelint$configIgnoreFiles = string | Array<string>;

export type stylelint$configRuleSettings = any | [any, Object];
export type stylelint$configRules = {
  [ruleName: string]: stylelint$configRuleSettings
};

export type stylelint$config = {
  extends?: stylelint$configExtends,
  plugins?: stylelint$configPlugins,
  pluginFunctions?: {
    [pluginName: string]: Function
  },
  processors?: stylelint$configProcessors,
  processorFunctions?: Array<Function>,
  ignoreFiles?: stylelint$configIgnoreFiles,
  ignorePatterns?: string,
  rules?: stylelint$configRules,
  codeProcessors?: Array<Function>,
  resultProcessors?: Array<Function>,
  quiet?: boolean
};

export type stylelint$syntaxes = "scss" | "less" | "sugarss";

export type stylelint$options = {
  config?: stylelint$config,
  configFile?: string,
  configBasedir?: string,
  configOverrides?: Object,
  ignoreDisables?: boolean,
  ignorePath?: string,
  reportNeedlessDisables?: boolean,
  syntax?: stylelint$syntaxes,
  customSyntax?: string,
  fix?: boolean
};

export type stylelint$internalApi = {
  _options: stylelint$options,
  _extendExplorer: { load: Function },
  _fullExplorer: { load: Function },
  _configCache: Map<string, Object>,
  _specifiedConfigCache: Map<string, Object>,
  _postcssResultCache: Map<string, Object>,

  _augmentConfig: Function,
  _getPostcssResult: Function,
  _lintSource: Function,
  _createStylelintResult: Function,
  _createEmptyPostcssResult?: Function,

  getConfigForFile: Function,
  isPathIgnored: Function,
  lintSource: Function
};

export type stylelint$warning = {
  line: number,
  column: number,
  rule: string,
  severity: string,
  text: string
};

export type stylelint$result = {
  source: string,
  deprecations: Array<{
    text: string,
    reference: string
  }>,
  invalidOptionWarnings: Array<{
    text: string
  }>,
  parseErrors: Array<stylelint$warning>,
  errored?: boolean,
  warnings: Array<stylelint$warning>,
  ignored?: boolean,
  _postcssResult?: Object
};

export type stylelint$needlessDisablesReport = Array<{
  source: string,
  ranges: Array<{
    source: string,
    ranges: Array<{
      start: number,
      end?: number
    }>
  }>
}>;

export type stylelint$standaloneReturnValue = {
  results: Array<stylelint$result>,
  errored: boolean,
  output: any,
  needlessDisables?: stylelint$needlessDisablesReport
};

export type stylelint$standaloneOptions = {
  files?: string | Array<string>,
  cache?: boolean,
  cacheLocation?: string,
  code?: string,
  codeFilename?: string,
  config?: stylelint$config,
  configFile?: string,
  configBasedir?: string,
  configOverrides?: Object,
  ignoreDisables?: boolean,
  ignorePath?: string,
  reportNeedlessDisables?: boolean,
  syntax?: stylelint$syntaxes,
  customSyntax?: string,
  formatter?: "json" | "string" | "verbose" | Function,
  disableDefaultIgnores?: boolean,
  allowEmptyInput?: boolean,
  fix?: boolean
};
