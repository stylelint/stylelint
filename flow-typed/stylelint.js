export type stylelint$configExtends = string | Array<string>;
export type stylelint$configPlugins = string | Array<string>;
export type stylelint$configProcessors =
  | string
  | Array<string | [string, Object]>;
export type stylelint$configIgnoreFiles = string | Array<string>;

export type stylelint$configRuleSettings = any | [any, Object];
export type stylelint$configRules = {
  [ruleName: string]: stylelint$configRuleSettings
};

export type stylelint$config = {
  extends?: stylelint$configExtends,
  plugins?: stylelint$configPlugins,
  pluginFunctions?: {
    [pluginName: string]: Function
  },
  processors?: stylelint$configProcessors,
  processorFunctions?: Array<Function>,
  ignoreFiles?: stylelint$configIgnoreFiles,
  ignorePatterns?: string,
  rules?: stylelint$configRules,
  codeProcessors?: Array<Function>,
  resultProcessors?: Array<Function>,
  quiet?: boolean,
  defaultSeverity?: string
};

export type stylelint$syntaxes = "scss" | "less" | "sugarss";

export type stylelint$options = {
  config?: stylelint$config,
  configFile?: string,
  configBasedir?: string,
  configOverrides?: Object,
  ignoreDisables?: boolean,
  ignorePath?: string,
  reportInvalidScopeDisables?: boolean,
  reportNeedlessDisables?: boolean,
  syntax?: stylelint$syntaxes,
  customSyntax?: string,
  fix?: boolean
};

export type stylelint$internalApi = {
  _options: stylelint$options,
  _extendExplorer: {
    search: string => Promise<null | Object>,
    load: string => Promise<null | Object>
  },
  _fullExplorer: {
    search: string => Promise<null | Object>,
    load: string => Promise<null | Object>
  },
  _configCache: Map<string, Object>,
  _specifiedConfigCache: Map<string, Object>,
  _postcssResultCache: Map<string, Object>,

  _augmentConfig: Function,
  _getPostcssResult: Function,
  _lintSource: Function,
  _createStylelintResult: Function,
  _createEmptyPostcssResult?: Function,

  getConfigForFile: Function,
  isPathIgnored: Function,
  lintSource: Function
};

export type stylelint$warning = {
  line: number,
  column: number,
  rule: string,
  severity: string,
  text: string
};

export type stylelint$result = {
  source: string,
  deprecations: Array<{
    text: string,
    reference: string
  }>,
  invalidOptionWarnings: Array<{
    text: string
  }>,
  parseErrors: Array<stylelint$warning>,
  errored?: boolean,
  warnings: Array<stylelint$warning>,
  ignored?: boolean,
  _postcssResult?: Object
};

export type stylelint$cssSyntaxError = {
  column: number,
  file?: string,
  input: {
    column: number,
    file?: string,
    line: number,
    source: string
  },
  line: number,
  message: string,
  name: string,
  reason: string,
  source: string
};

export type stylelint$disableOptionsReport = Array<{
  source: string,
  ranges: Array<{
    unusedRule: string,
    start: number,
    end?: number
  }>
}>;

export type stylelint$standaloneReturnValue = {
  results: Array<stylelint$result>,
  errored: boolean,
  output: any,
  maxWarningsExceeded?: {
    maxWarnings: number,
    foundWarnings: number
  },
  needlessDisables?: stylelint$disableOptionsReport,
  invalidScopeDisables?: stylelint$disableOptionsReport
};

export type stylelint$standaloneOptions = {
  files?: string | Array<string>,
  globbyOptions?: Object,
  cache?: boolean,
  cacheLocation?: string,
  code?: string,
  codeFilename?: string,
  config?: stylelint$config,
  configFile?: string,
  configBasedir?: string,
  configOverrides?: Object,
  printConfig?: string,
  ignoreDisables?: boolean,
  ignorePath?: string,
  ignorePattern?: RegExp,
  reportNeedlessDisables?: boolean,
  reportInvalidScopeDisables?: boolean,
  maxWarnings?: number,
  syntax?: stylelint$syntaxes,
  customSyntax?: string,
  formatter?: "compact" | "json" | "string" | "unix" | "verbose" | Function,
  disableDefaultIgnores?: boolean,
  fix?: boolean,
  allowEmptyInput: boolean
};
