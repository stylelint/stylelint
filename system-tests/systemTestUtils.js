"use strict";

const _ = require("lodash");
const path = require("path");
const stripAnsi = require("strip-ansi");

function caseFilePath(caseNumber, fileName) {
  return path.join(__dirname, caseNumber, fileName);
}

function caseStylesheetGlob(caseNumber) {
  return caseFilePath(caseNumber, "stylesheet.*");
}

function caseConfig(caseNumber, ext) {
  ext = ext || "json";

  return caseFilePath(caseNumber, `config.${ext}`);
}

function prepResults(results) {
  return results.map(result => {
    // The _postcssResult object is not part of our API and is huge
    const preppedResult = _.omit(result, "_postcssResult");

    // The `source` of each file will not be the same on different machines or platforms
    preppedResult.source = path
      .relative(__dirname, result.source)
      .replace(new RegExp(`\\${path.sep}`, "g"), "/");

    return preppedResult;
  });
}

function stripColors(input) {
  return stripAnsi(input);
}

module.exports = {
  caseFilePath,
  caseStylesheetGlob,
  caseConfig,
  prepResults,
  stripColors
};
