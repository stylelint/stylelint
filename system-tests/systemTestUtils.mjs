import { copyFile, readFile } from 'node:fs/promises';
import { fileURLToPath } from 'node:url';
import os from 'node:os';
import path from 'node:path';
import process from 'node:process';

import replaceBackslashes from '../lib/testUtils/replaceBackslashes.mjs';
import uniqueId from '../lib/testUtils/uniqueId.mjs';

const __dirname = fileURLToPath(new URL('.', import.meta.url));

export function caseFilePath(caseNumber, fileName = 'stylesheet') {
	return replaceBackslashes(path.join(__dirname, caseNumber, fileName));
}

export function caseFiles(caseNumber) {
	return caseFilePath(caseNumber, 'stylesheet.*');
}

export function caseConfigFile(caseNumber, ext = 'json') {
	return caseFilePath(caseNumber, `config.${ext}`);
}

export async function caseConfig(caseNumber) {
	return JSON.parse(await readFile(caseConfigFile(caseNumber, 'json'), 'utf8'));
}

export async function caseCode(caseNumber, ext = 'css') {
	return readFile(`${caseFilePath(caseNumber)}.${ext}`, 'utf8');
}

export async function caseFilesForFix(caseNumber, ext = 'css') {
	const tempPath = replaceBackslashes(path.join(os.tmpdir(), `stylesheet-${uniqueId()}.${ext}`));

	await copyFile(path.join(__dirname, caseNumber, `stylesheet.${ext}`), tempPath);

	return tempPath;
}

export function prepForSnapshot({ results, cwd, output, report, ...rest }) {
	// If output isn't fixed code
	if (output.startsWith('[')) {
		// The `source` of each file varies between platforms or if a tmp file is used
		output = JSON.parse(output).map((warning) => ({
			...warning,
			source: '/path/to/dummy.css',
		}));
	}

	if (report) {
		// The `source` of each file varies between platforms or if a tmp file is used
		report = JSON.parse(report).map((warning) => ({
			...warning,
			source: '/path/to/dummy.css',
		}));
	}

	return {
		cwd: path.relative(process.cwd(), cwd),
		// The _postcssResult object is not part of our API and is huge
		results: results.map((result) => {
			delete result._postcssResult;

			return { ...result, source: '/path/to/dummy.css' };
		}),
		output, // TODO: Deprecated. Remove in the next major version.
		report,
		...rest,
	};
}
