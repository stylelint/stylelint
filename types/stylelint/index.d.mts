export {
	default,
	BuiltInRules,
	Config,
	ConfigRuleSettings,
	CosmiconfigResult,
	CssSyntaxError,
	CustomSyntax,
	DisableOptions,
	DisableOptionsReport,
	DisablePropertyName,
	DisableReportRange,
	DisabledRange,
	DisabledRangeObject,
	DisabledWarning,
	Formatter,
	FormatterType,
	Formatters,
	GetLintSourceOptions,
	GetPostcssOptions,
	InternalApi,
	LintResult,
	LinterOptions,
	LinterResult,
	LonghandSubPropertiesOfShorthandProperties,
	Plugin,
	PostcssPluginOptions,
	PostcssResult,
	Problem,
	PublicApi,
	Rule,
	RuleBase,
	RuleContext,
	RuleMessage,
	RuleMessages,
	RuleMeta,
	RuleOptions,
	RuleOptionsPossible,
	RuleSeverity,
	Severity,
	ShorthandProperties,
	StylelintPostcssResult,
	Utils,
	Warning,
	WarningOptions,
} from './index.js';
