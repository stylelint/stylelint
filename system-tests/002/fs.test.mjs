import postcssScss from 'postcss-scss';

import stylelint from '../../lib/index.mjs';

import { caseConfigFile, caseFiles, prepForSnapshot } from '../systemTestUtils.mjs';

const CASE_NUMBER = '002';

it('fs - invalid twbs buttons and their config', async () => {
	expect(
		prepForSnapshot(
			await stylelint.lint({
				files: caseFiles(CASE_NUMBER),
				configFile: caseConfigFile(CASE_NUMBER),
				customSyntax: postcssScss,
			}),
		),
	).toMatchSnapshot();
}, 10000);
