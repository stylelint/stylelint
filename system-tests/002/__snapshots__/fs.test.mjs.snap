// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fs - invalid twbs buttons and their config 1`] = `
{
  "cwd": "",
  "errored": true,
  "output": [
    {
      "deprecations": [
        {
          "text": "The "at-rule-semicolon-space-before" rule is deprecated.",
        },
        {
          "text": "The "at-rule-name-space-after" rule is deprecated.",
        },
        {
          "text": "The "declaration-block-semicolon-newline-before" rule is deprecated.",
        },
        {
          "text": "The "max-empty-lines" rule is deprecated.",
        },
        {
          "text": "The "number-leading-zero" rule is deprecated.",
        },
        {
          "text": "The "selector-list-comma-newline-before" rule is deprecated.",
        },
        {
          "text": "The "selector-list-comma-space-after" rule is deprecated.",
        },
        {
          "text": "The "selector-list-comma-space-before" rule is deprecated.",
        },
        {
          "text": "The "string-quotes" rule is deprecated.",
        },
        {
          "text": "The "unicode-bom" rule is deprecated.",
        },
        {
          "text": "The "value-list-comma-newline-after" rule is deprecated.",
        },
        {
          "text": "The "value-list-comma-newline-before" rule is deprecated.",
        },
        {
          "text": "The "value-list-comma-space-after" rule is deprecated.",
        },
      ],
      "errored": true,
      "invalidOptionWarnings": [],
      "parseErrors": [],
      "warnings": [
        {
          "column": 10,
          "endColumn": 15,
          "endLine": 79,
          "line": 79,
          "rule": "color-named",
          "severity": "error",
          "text": "Unexpected named color "white" (color-named)",
        },
        {
          "column": 1,
          "endColumn": 12,
          "endLine": 118,
          "line": 118,
          "rule": "selector-no-qualifying-type",
          "severity": "error",
          "text": "Unexpected qualifying type selector "a.btn-block" (selector-no-qualifying-type)",
        },
        {
          "column": 3,
          "endColumn": 4,
          "endLine": 125,
          "line": 123,
          "rule": "selector-no-qualifying-type",
          "severity": "error",
          "text": "Unexpected qualifying type selector "a.btn-block" (selector-no-qualifying-type)",
        },
      ],
    },
  ],
  "reportedDisables": [],
  "results": [
    {
      "deprecations": [
        {
          "reference": undefined,
          "text": "The "at-rule-semicolon-space-before" rule is deprecated.",
        },
        {
          "reference": undefined,
          "text": "The "at-rule-name-space-after" rule is deprecated.",
        },
        {
          "reference": undefined,
          "text": "The "declaration-block-semicolon-newline-before" rule is deprecated.",
        },
        {
          "reference": undefined,
          "text": "The "max-empty-lines" rule is deprecated.",
        },
        {
          "reference": undefined,
          "text": "The "number-leading-zero" rule is deprecated.",
        },
        {
          "reference": undefined,
          "text": "The "selector-list-comma-newline-before" rule is deprecated.",
        },
        {
          "reference": undefined,
          "text": "The "selector-list-comma-space-after" rule is deprecated.",
        },
        {
          "reference": undefined,
          "text": "The "selector-list-comma-space-before" rule is deprecated.",
        },
        {
          "reference": undefined,
          "text": "The "string-quotes" rule is deprecated.",
        },
        {
          "reference": undefined,
          "text": "The "unicode-bom" rule is deprecated.",
        },
        {
          "reference": undefined,
          "text": "The "value-list-comma-newline-after" rule is deprecated.",
        },
        {
          "reference": undefined,
          "text": "The "value-list-comma-newline-before" rule is deprecated.",
        },
        {
          "reference": undefined,
          "text": "The "value-list-comma-space-after" rule is deprecated.",
        },
      ],
      "errored": true,
      "ignored": undefined,
      "invalidOptionWarnings": [],
      "parseErrors": [],
      "warnings": [
        {
          "column": 10,
          "endColumn": 15,
          "endLine": 79,
          "line": 79,
          "rule": "color-named",
          "severity": "error",
          "text": "Unexpected named color "white" (color-named)",
        },
        {
          "column": 1,
          "endColumn": 12,
          "endLine": 118,
          "line": 118,
          "rule": "selector-no-qualifying-type",
          "severity": "error",
          "text": "Unexpected qualifying type selector "a.btn-block" (selector-no-qualifying-type)",
        },
        {
          "column": 3,
          "endColumn": 4,
          "endLine": 125,
          "line": 123,
          "rule": "selector-no-qualifying-type",
          "severity": "error",
          "text": "Unexpected qualifying type selector "a.btn-block" (selector-no-qualifying-type)",
        },
      ],
    },
  ],
  "ruleMetadata": {
    "at-rule-name-space-after": {
      "deprecated": true,
      "fixable": true,
      "url": "https://stylelint.io/user-guide/rules/at-rule-name-space-after",
    },
    "at-rule-no-vendor-prefix": {
      "fixable": true,
      "url": "https://stylelint.io/user-guide/rules/at-rule-no-vendor-prefix",
    },
    "at-rule-semicolon-space-before": {
      "deprecated": true,
      "url": "https://stylelint.io/user-guide/rules/at-rule-semicolon-space-before",
    },
    "color-named": {
      "url": "https://stylelint.io/user-guide/rules/color-named",
    },
    "declaration-block-semicolon-newline-before": {
      "deprecated": true,
      "url": "https://stylelint.io/user-guide/rules/declaration-block-semicolon-newline-before",
    },
    "declaration-no-important": {
      "url": "https://stylelint.io/user-guide/rules/declaration-no-important",
    },
    "font-family-name-quotes": {
      "fixable": true,
      "url": "https://stylelint.io/user-guide/rules/font-family-name-quotes",
    },
    "font-weight-notation": {
      "fixable": true,
      "url": "https://stylelint.io/user-guide/rules/font-weight-notation",
    },
    "function-url-no-scheme-relative": {
      "url": "https://stylelint.io/user-guide/rules/function-url-no-scheme-relative",
    },
    "function-url-quotes": {
      "fixable": true,
      "url": "https://stylelint.io/user-guide/rules/function-url-quotes",
    },
    "max-empty-lines": {
      "deprecated": true,
      "fixable": true,
      "url": "https://stylelint.io/user-guide/rules/max-empty-lines",
    },
    "media-feature-name-no-vendor-prefix": {
      "fixable": true,
      "url": "https://stylelint.io/user-guide/rules/media-feature-name-no-vendor-prefix",
    },
    "number-leading-zero": {
      "deprecated": true,
      "fixable": true,
      "url": "https://stylelint.io/user-guide/rules/number-leading-zero",
    },
    "property-no-vendor-prefix": {
      "fixable": true,
      "url": "https://stylelint.io/user-guide/rules/property-no-vendor-prefix",
    },
    "selector-attribute-quotes": {
      "fixable": true,
      "url": "https://stylelint.io/user-guide/rules/selector-attribute-quotes",
    },
    "selector-class-pattern": {
      "url": "https://stylelint.io/user-guide/rules/selector-class-pattern",
    },
    "selector-list-comma-newline-before": {
      "deprecated": true,
      "fixable": true,
      "url": "https://stylelint.io/user-guide/rules/selector-list-comma-newline-before",
    },
    "selector-list-comma-space-after": {
      "deprecated": true,
      "fixable": true,
      "url": "https://stylelint.io/user-guide/rules/selector-list-comma-space-after",
    },
    "selector-list-comma-space-before": {
      "deprecated": true,
      "fixable": true,
      "url": "https://stylelint.io/user-guide/rules/selector-list-comma-space-before",
    },
    "selector-max-attribute": {
      "url": "https://stylelint.io/user-guide/rules/selector-max-attribute",
    },
    "selector-max-class": {
      "url": "https://stylelint.io/user-guide/rules/selector-max-class",
    },
    "selector-max-combinators": {
      "url": "https://stylelint.io/user-guide/rules/selector-max-combinators",
    },
    "selector-max-compound-selectors": {
      "url": "https://stylelint.io/user-guide/rules/selector-max-compound-selectors",
    },
    "selector-max-id": {
      "url": "https://stylelint.io/user-guide/rules/selector-max-id",
    },
    "selector-max-type": {
      "url": "https://stylelint.io/user-guide/rules/selector-max-type",
    },
    "selector-max-universal": {
      "url": "https://stylelint.io/user-guide/rules/selector-max-universal",
    },
    "selector-no-qualifying-type": {
      "url": "https://stylelint.io/user-guide/rules/selector-no-qualifying-type",
    },
    "selector-no-vendor-prefix": {
      "fixable": true,
      "url": "https://stylelint.io/user-guide/rules/selector-no-vendor-prefix",
    },
    "shorthand-property-no-redundant-values": {
      "fixable": true,
      "url": "https://stylelint.io/user-guide/rules/shorthand-property-no-redundant-values",
    },
    "string-quotes": {
      "deprecated": true,
      "fixable": true,
      "url": "https://stylelint.io/user-guide/rules/string-quotes",
    },
    "unicode-bom": {
      "deprecated": true,
      "url": "https://stylelint.io/user-guide/rules/unicode-bom",
    },
    "value-list-comma-newline-after": {
      "deprecated": true,
      "fixable": true,
      "url": "https://stylelint.io/user-guide/rules/value-list-comma-newline-after",
    },
    "value-list-comma-newline-before": {
      "deprecated": true,
      "url": "https://stylelint.io/user-guide/rules/value-list-comma-newline-before",
    },
    "value-list-comma-space-after": {
      "deprecated": true,
      "fixable": true,
      "url": "https://stylelint.io/user-guide/rules/value-list-comma-space-after",
    },
    "value-no-vendor-prefix": {
      "fixable": true,
      "url": "https://stylelint.io/user-guide/rules/value-no-vendor-prefix",
    },
  },
}
`;
